app/src/test/java/org/schabi/newpipe/player/playqueue/PlayQueueTest.java

```diff
@@ -0,0 +1,200 @@
+            nonEmptyQueue = spy(mockPlayQueue(
+                    0, Collections.nCopies(SIZE, mock(PlayQueueItem.class))
+            ));
+            emptyQueue = spy(mockPlayQueue(0, new ArrayList<>()));
+        }
+
+        @Test
+        public void negative() {
+            nonEmptyQueue.setIndex(-5);
+            assertEquals(0, nonEmptyQueue.getIndex());
+
+            emptyQueue.setIndex(-5);
+            assertEquals(0, nonEmptyQueue.getIndex());
+        }
+
+        @Test
+        public void inBounds() {
+            nonEmptyQueue.setIndex(2);
+            assertEquals(2, nonEmptyQueue.getIndex());
+
+            // emptyQueue not tested because 0 isn't technically inBounds
+        }
+
+        @Test
+        public void outOfBoundIsComplete() {
+            doReturn(true).when(nonEmptyQueue).isComplete();
+            nonEmptyQueue.setIndex(7);
+            assertEquals(2, nonEmptyQueue.getIndex());
+
+            doReturn(true).when(emptyQueue).isComplete();
+            emptyQueue.setIndex(2);
+            assertEquals(0, emptyQueue.getIndex());
+        }
+
+        @Test
+        public void outOfBoundsNotComplete() {
+            doReturn(false).when(nonEmptyQueue).isComplete();
+            nonEmptyQueue.setIndex(7);
+            assertEquals(SIZE - 1, nonEmptyQueue.getIndex());
+
+            doReturn(false).when(emptyQueue).isComplete();
+            emptyQueue.setIndex(2);
+            assertEquals(0, emptyQueue.getIndex());
+        }
+
+        @Test
+        public void indexZero() {
+            nonEmptyQueue.setIndex(0);
+            assertEquals(0, nonEmptyQueue.getIndex());
+
+            doReturn(true).when(emptyQueue).isComplete();
+            emptyQueue.setIndex(0);
+            assertEquals(0, emptyQueue.getIndex());
+
+            doReturn(false).when(emptyQueue).isComplete();
+            emptyQueue.setIndex(0);
+            assertEquals(0, emptyQueue.getIndex());
+        }
+
+        @SuppressWarnings("unchecked")
+        @Test
+        public void addToHistory() throws NoSuchFieldException, IllegalAccessException {
+            final Field field;
+            field = PlayQueue.class.getDeclaredField("history");
+            field.setAccessible(true);
+            List<PlayQueueItem> history;
+
+            /*
+            history's size is currently 1. 0 is the also the current index, so history should not
+            be affected.
+             */
+            nonEmptyQueue.setIndex(0);
+            history = (List<PlayQueueItem>) Objects.requireNonNull(
+                    field.get(nonEmptyQueue)
+            );
+            assertEquals(1, history.size());
+
+            // Index 3 != 0, so the second history element should be the item at streams[3]
+            nonEmptyQueue.setIndex(3);
+            history = (List<PlayQueueItem>) Objects.requireNonNull(
+                    field.get(nonEmptyQueue)
+            );
+            assertEquals(nonEmptyQueue.getItem(3), history.get(1));
```

Comment:
I have trouble understanding what this does and why reflection is necessary for it. Can't this somehow be done with mocks?

Reply:
Adding to history is a possible side effect of `setIndex(int): void`, so I thought I’d test that. I haven’t really worked with Mockito, so I just did this the old-fashioned way. `Whitebox.getInternalState(…)` also requires PowerMock.

Reply:
I would not test this or at least not in this way. This is internal state of the class which should have some impact on the result of other methods. These other methods should be called and asserted on